File: android/app/src/main/java/com/gradethread/app/billing/CreditTopUpFlow.kt
package com.gradethread.app.billing

import com.gradethread.app.platform.net.Backoff

/**
 * US-1338: what happens between "the buyer paid" and "the submit button
 * unblocks" (iOS `CreditTopUpFlow`, US-809).
 *
 * Pure over injected seams — the balance read and the sleep are both passed in
 * — so the whole machine unit-tests with no Play Store and no network.
 *
 * It never submits anything. It only unblocks the submit button that was
 * already there, so there is no double-grant risk here; the server's grant RPC
 * is independently idempotent by purchase token.
 */
object CreditTopUpFlow {

    /**
     * Bounded so a grant that never lands still returns control. 12 attempts at
     * 1,2,4,8,8,… is ~78s — long enough for Play's own propagation, short
     * enough that a seller isn't watching a spinner indefinitely.
     */
    const val MAX_POLLS = 12

    sealed class State {
        object Idle : State()

        /** Play's purchase dialog is up. */
        object Purchasing : State()

        /** Paid; waiting for the credits to appear on the account. */
        object AwaitingGrant : State()

        data class Granted(val balance: Int) : State()

        /**
         * The grant didn't appear in the window. NOT an error state — the sheet
         * stays usable and offers "Check again", because the overwhelmingly
         * likely cause is propagation lag, not a lost purchase.
         */
        object TimedOut : State()

        data class Failed(val message: String) : State()
    }

    /**
     * Did the grant land?
     *
     * Detected by a STRICT INCREASE over the pre-purchase balance rather than
     * by an absolute target, which is what makes it tolerant of a transient
     * read failure (null just retries) and of a second grant arriving from
     * another device mid-poll.
     */
    fun granted(baseline: Int, balance: Int?): Boolean = balance != null && balance > baseline

    /** Backoff between polls; mirrors the grading poll's curve. */
    fun delayFor(attempt: Int): Long = Backoff.delayMillis(attempt, baseMillis = 1_000, capMillis = 8_000)

    /**
     * Poll until the balance rises above [baseline] or the window elapses.
     *
     * @param verifiedBalance the balance the verify call itself reported, if
     *   any. The Android verify endpoint applies the grant BEFORE it responds,
     *   so this is usually authoritative immediately and no polling happens at
     *   all — unlike iOS, where StoreKit confirms first and the grant lands out
     *   of band. The poll remains as the fallback for a response we didn't get
     *   to read (dropped connection after the server committed).
     * @return the terminal state; [State.TimedOut] still means "re-validate",
     *   never "the purchase failed".
     */
    suspend fun awaitGrant(
        baseline: Int,
        verifiedBalance: Int?,
        fetchBalance: suspend () -> Int?,
        sleep: suspend (Long) -> Unit,
        maxPolls: Int = MAX_POLLS,
    ): State {
        if (granted(baseline, verifiedBalance)) return State.Granted(verifiedBalance!!)

        for (attempt in 0 until maxPolls) {
            val balance = fetchBalance()
            if (granted(baseline, balance)) return State.Granted(balance!!)
            // No sleep after the final attempt — we're about to give up anyway.
            if (attempt < maxPolls - 1) sleep(delayFor(attempt))
        }
        return State.TimedOut
    }
}

File: ios/GradeThread/Grading/CreditTopUpFlow.swift
import Foundation

/// Poll/re-validate state machine for an in-flow grade-credit top-up (US-809).
///
/// When a seller is blocked from grading (no included grades left + a zero
/// credit balance) they can buy a credit pack right inside the grading sheet.
/// The grant lands asynchronously — StoreKit confirms the purchase, then
/// `/appstore/verify` calls the idempotent `grant_appstore_credits` RPC — so
/// the user's `grade_credit_balance` lags the purchase by a moment. This
/// machine polls the billing snapshot with bounded retries + exponential
/// backoff until the granted credits appear, then asks the caller to
/// re-validate so the submit button unblocks. On timeout it still re-validates
/// (the grant may have landed between the final poll and now) and leaves the
/// sheet usable with a "Check again" retry affordance.
///
/// It never submits anything — it only unblocks the existing submit button —
/// so there is no double-grant risk here (and the server grant RPC is itself
/// idempotent by transaction id).
///
/// Pure of UI + network: the balance fetch and the sleep are injected, and the
/// re-validate is passed per call, so the whole machine is unit-tested with no
/// Supabase/StoreKit dependency.
@MainActor
@Observable
final class CreditTopUpFlow {

    enum State: Equatable {
        /// No top-up in progress.
        case idle
        /// A purchase succeeded; polling the billing snapshot for the grant.
        case awaitingGrant
        /// The grant landed — the associated value is the new credit balance.
        case granted(Int)
        /// The grant didn't appear within the poll window. The sheet stays
        /// usable; the user can poll again ("Check again").
        case timedOut
    }

    private(set) var state: State = .idle

    /// Current credit balance (server truth), or nil on a transient fetch error.
    private let fetchBalance: () async -> Int?
    /// Injectable sleep (no-op in tests) — receives the backoff delay in nanos.
    private let sleep: (UInt64) async -> Void
    /// Max poll attempts before giving up (exponential backoff capped at ~8s).
    let maxPolls: Int

    init(
        maxPolls: Int = 12,
        fetchBalance: @escaping () async -> Int?,
        sleep: @escaping (UInt64) async -> Void = { try? await Task.sleep(nanoseconds: $0) }
    ) {
        self.maxPolls = maxPolls
        self.fetchBalance = fetchBalance
        self.sleep = sleep
    }

    /// Poll until the credit balance exceeds `baseline` (the grant landed) or
    /// the window elapses, then run `revalidate`. The grant is detected by a
    /// strict increase over the pre-purchase balance, which tolerates transient
    /// nil fetches (a failed poll just retries on the next tick).
    func awaitGrant(baseline: Int, revalidate: () async -> Void) async {
        state = .awaitingGrant
        for attempt in 0..<maxPolls {
            if Task.isCancelled { break }
            if let balance = await fetchBalance(), balance > baseline {
                state = .granted(balance)
                await revalidate()
                return
            }
            // No sleep after the final attempt — we're about to give up.
            if attempt < maxPolls - 1 {
                await sleep(Backoff.jitteredDelayNanos(attempt: attempt, base: 1, cap: 8))
            }
        }
        state = .timedOut
        // The grant may have arrived between the last poll and now; re-validate
        // so the server (source of truth) decides whether submit is unblocked.
        await revalidate()
    }

    /// Return to idle (e.g. when a fresh validation cycle starts).
    func reset() { state = .idle }
}
